compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2466,11 +2466,26 @@ impl<'a> Parser<'a> {
         if !self.eat_keyword(kw::In) {
             self.error_missing_in_for_loop();
         }
+        let expr_span = self.token.span;
         self.check_for_for_in_in_typo(self.prev_token.span);
         let expr = self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None)?;
 
         let pat = self.recover_parens_around_for_head(pat, begin_paren);
 
+        // Recover from missing expression in `for` loop
+        if matches!(expr.kind, ExprKind::Block(..))
+            && !matches!(self.token.kind, token::OpenDelim(token::Delimiter::Brace))
+            && self.may_recover()
+        {
+            let mut diag = self.sess.struct_err("expected expression, found block");
```

Comment:
You still didn't migrate this to a diagnostic derive :sweat_smile:, it just went from `struct_span_err` to `struct_err`?

You need to make a struct, adding `derive(Diagnostic)` and adding the corresponding entries to the `parser.ftl` file. There should be lots of example in the git history for https://github.com/rust-lang/rust/commits/master/compiler/rustc_parse/src/errors.rs